### PR TITLE
Don't reconcile roles and policies if a profile is provided

### PR DESCRIPTION
**Pull request by olemarkus**
If we are using a shared IAM profile, kops shouldn't try to do anything with policies and profile role.

### CODE REVIEW

File: pkg/model/awsmodel/iam.go [hunk 1]
```diff
@@ -333,10 +333,7 @@ func (b *IAMModelBuilder) buildIAMTasks(role iam.Subject, iamName string, c *fi.
 			}
 
 			c.AddTask(t)
-		}
 
-		// Generate additional policies if needed, and attach to existing role
-		if !shared {
```
Comment (hakman) [inline pkg/model/awsmodel/iam.go:339]: Does it hurt in any way to leave this as 2 separate blocks of code?

Comment (olemarkus) [reply pkg/model/awsmodel/iam.go:339]: `iamRole` is defined in one and used in the other.

Comment (hakman) [reply pkg/model/awsmodel/iam.go:339]: Maybe just create 2 blocks and keep the original comments for each block?

Comment (olemarkus) [reply pkg/model/awsmodel/iam.go:339]: yeah removing the comments were unecessary.

File: pkg/model/awsmodel/iam.go [hunk 2]
```diff
@@ -297,24 +288,34 @@ func (b *IAMModelBuilder) buildIAMTasks(role iam.Subject, iamName string, c *fi.
...
+				Tags:      b.CloudTags(iamName, shared),
 			}
 			c.AddTask(iamInstanceProfile)
 		}
 
-		{
-			iamInstanceProfileRole := &awstasks.IAMInstanceProfileRole{
-				Name:      fi.String(iamName),
-				Lifecycle: b.Lifecycle,
+		// Create External Policy tasks
```
Comment (hakman) [inline pkg/model/awsmodel/iam.go:296]: This is duplicated now.
```suggestion
```

File: pkg/model/awsmodel/iam.go [hunk 3]
```diff
@@ -333,10 +334,8 @@ func (b *IAMModelBuilder) buildIAMTasks(role iam.Subject, iamName string, c *fi.
 			}
 
 			c.AddTask(t)
-		}
 
-		// Generate additional policies if needed, and attach to existing role
-		if !shared {
+			// Generate additional policies if needed, and attach to existing role
```
Comment (hakman) [inline pkg/model/awsmodel/iam.go:338]: Having this as a block would make it more readable, but your choice.
```suggestion
			// Generate additional policies if needed, and attach to existing role
			{
```

File: pkg/model/awsmodel/iam.go [hunk 4]
```diff
@@ -297,24 +288,34 @@ func (b *IAMModelBuilder) buildIAMTasks(role iam.Subject, iamName string, c *fi.
...
+					Name:      fi.String(iamName),
+					Lifecycle: b.Lifecycle,
+
+					InstanceProfile: iamInstanceProfile,
+					Role:            iamRole,
+				}
+				c.AddTask(iamInstanceProfileRole)
+			}
+
+			// Create External Policy tasks
```
Comment (hakman) [inline pkg/model/awsmodel/iam.go:318]: Having this as a block would make it more readable, but your choice.
```suggestion
			// Create External Policy tasks
			{
```

**Review by hakman — CHANGES_REQUESTED** [review]
A few more nits, will finish the review once they are addressed.

**Review by hakman — APPROVED** [review]

**Pull request merged**